pq-crypto/sike_r3/sikep434r3_fp.c

```diff
@@ -0,0 +1,256 @@
+    c[1] = temp & mask_low;                   // C10
+    carry = temp & mask_high; 
+    c[1] ^= (ahbh & mask_high) + carry;       // C11
+}
+
+// Multiprecision comba multiply, c = a*b, where lng(a) = lng(b) = nwords.
+void mp_mul(const digit_t* a, const digit_t* b, digit_t* c, const unsigned int nwords)
+{
+  unsigned int i, j;
+  digit_t t = 0, u = 0, v = 0, UV[2];
+  unsigned int carry;
+
+  for (i = 0; i < nwords; i++) {
+      for (j = 0; j <= i; j++) {
+          SIKE_P434_R3_MUL(a[j], b[i-j], UV+1, UV[0]);
+          SIKE_P434_R3_ADDC(0, UV[0], v, carry, v);
+          SIKE_P434_R3_ADDC(carry, UV[1], u, carry, u);
+          t += carry;
+      }
+      c[i] = v;
+      v = u;
+      u = t;
+      t = 0;
+  }
+
+  for (i = nwords; i < 2*nwords-1; i++) {
+      for (j = i-nwords+1; j < nwords; j++) {
+          SIKE_P434_R3_MUL(a[j], b[i-j], UV+1, UV[0]);
+          SIKE_P434_R3_ADDC(0, UV[0], v, carry, v);
+          SIKE_P434_R3_ADDC(carry, UV[1], u, carry, u);
+          t += carry;
+      }
+      c[i] = v;
+      v = u;
+      u = t;
+      t = 0;
+  }
+  c[2*nwords-1] = v;
+}
+
+// Efficient Montgomery reduction using comba and exploiting the special form of the prime p434.
+// mc = ma*R^-1 mod p434x2, where R = 2^448.
+// If ma < 2^448*p434, the output mc is in the range [0, 2*p434-1].
+// ma is assumed to be in Montgomery representation.
+void rdc_mont(digit_t* ma, digit_t* mc)
+{
+  unsigned int i, j, carry, count = SIKE_P434_R3_ZERO_WORDS;
+  digit_t UV[2], t = 0, u = 0, v = 0;
+
+  for (i = 0; i < SIKE_P434_R3_NWORDS_FIELD; i++) {
+      mc[i] = 0;
+  }
+
+  for (i = 0; i < SIKE_P434_R3_NWORDS_FIELD; i++) {
+      for (j = 0; j < i; j++) {
+          if (j < (i-SIKE_P434_R3_ZERO_WORDS+1)) {
+              SIKE_P434_R3_MUL(mc[j], ((const digit_t*)p434p1)[i-j], UV+1, UV[0]);
+              SIKE_P434_R3_ADDC(0, UV[0], v, carry, v);
+              SIKE_P434_R3_ADDC(carry, UV[1], u, carry, u);
+              t += carry;
+          }
+      }
+      SIKE_P434_R3_ADDC(0, v, ma[i], carry, v);
+      SIKE_P434_R3_ADDC(carry, u, 0, carry, u);
+      t += carry;
+      mc[i] = v;
+      v = u;
+      u = t;
+      t = 0;
+  }
+
+  for (i = SIKE_P434_R3_NWORDS_FIELD; i < 2*SIKE_P434_R3_NWORDS_FIELD-1; i++) {
+      if (count > 0) {
+          count -= 1;
+      }
+      for (j = i-SIKE_P434_R3_NWORDS_FIELD+1; j < SIKE_P434_R3_NWORDS_FIELD; j++) {
+          if (j < (SIKE_P434_R3_NWORDS_FIELD-count)) {
+              SIKE_P434_R3_MUL(mc[j], ((const digit_t*)p434p1)[i-j], UV+1, UV[0]);
+              SIKE_P434_R3_ADDC(0, UV[0], v, carry, v);
+              SIKE_P434_R3_ADDC(carry, UV[1], u, carry, u);
+              t += carry;
+          }
+      }
+      SIKE_P434_R3_ADDC(0, v, ma[i], carry, v);
+      SIKE_P434_R3_ADDC(carry, u, 0, carry, u);
+      t += carry;
+      mc[i-SIKE_P434_R3_NWORDS_FIELD] = v;
+      v = u;
+      u = t;
+      t = 0;
+  }
+
+  // `carry` isn't read after this, but it's still a necessary argument to the macro
```

Comment:
You can delete this comment. I think it's leftover from the cppcheck lines that you previously removed.